### ISSUE TITLE
Decompiling a 2 MB dll eats up 32 Gigs of memory

### ISSUE THREAD

**Issue by invokr** [avast-tl/retdec#13]
The file I'm decompiling is kinda publicly available but I'm not sure how to share it with devs because the people who have written the code probably don't want it to be decompiled in the first place.

I'm running: ./decompiler.sh <path/to/dll>
Which at some point runs: `llvmir2hll -target-hll=c -var-renamer=readable -var-name-gen=fruit -var-name-gen-prefix= -call-info-obtainer=optim -arithm-expr-evaluator=c -validate-module -llvmir2bir-converter=orig -o [...].dll.c [...].dll.c.backend.bc -enable-debug -emit-debug-comments -config-path=[...].dll.c.json`

Once it reaches the step below, it slowly starts eating memory until at one point it quickly goes from 8GB -> 32 GB and then get's killed by docker.

How would I help debugging this? My guess would be it's running into some infinite depth loop.

```
Running phase: optimizations [normal] ( 52.09s )
 -> running RemoveUselessCastsOptimizer ( 52.09s )
 -> running UnusedGlobalVarOptimizer ( 52.63s )
 -> running DeadLocalAssignOptimizer ( 53.80s )
 -> running SimpleCopyPropagationOptimizer ( 65.16s )
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `v1_1002017b = (v3_10020166 + 1)` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `v1_1002031f = (v3_1002030a + 1)` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `v1_100200c9 = (v3_100200b4 + 1)` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `v1_10020f22 = (v3_10020f0d + 1)` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `goto 0x10020eca` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `eax_global_2d_to_2d_local = v3_10021abe` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `eax_global_2d_to_2d_local = v3_100224d9` -> skipping this edge
Warning: [NonRecursiveCFGBuilder] there is no node for an edge to `v10_1002391d = v2_1002386d` -> skipping this edge
```

**Comment by tpruzina** [comment]
Can you upload your dll somewhere? Looks interesting.

**Comment by invokr** [comment]
https://www.file-upload.net/download-12866451/2mbTurnsInto32Gigs.dll.html

Def. worth looking into in terms of denial of service. I've renamed the original dll, like I said, this isn't from an open source project or anything so I'm not sure regarding the legality of decompiling it.

**Comment by s3rvac** [comment]
Thank you for the report. We will investigate it. To limit virtual memory on Linux, you can run e.g. `ulimit -Sv 8388608` (8 GB limit) prior to running `decompile.sh`.

**Comment by ExtReMLapin** [comment]
# fucking hell

![](https://i.imgur.com/zXXNY8z.png)

**Comment by humblepride** [comment]
I have the same problem, is there any fix for this for now?

**Comment by Manouchehri** [comment]
@s3rvac Do your servers at work have 128 GB of RAM? I'm wondering if this is a larger design issue; formerly an analyst would be using RetDec on a remote server via the API, while with the open source release most analysts are running directly on their own low powered desktops/laptops.

That being said, somebody (or me eventually) should probably generate a couple flame graphs to check if there's any easy problems to squash.

**Comment by invokr** [comment]
With the testing I did yesterday I can already say that turning off optimizations (`-no-opts`) prevents this bug from occurring. Seeing that the optimizers running are printed, the bug would be most likely in the `SimpleCopyPropagationOptimizer`.

**Comment by s3rvac** [comment]
We are aware of the memory/speed issues when decompiling certain binaries. In most cases, they are caused either by invalid instruction decoding (when e.g. data are decompiled as code) or simply by the size of the code that is being decompiled (the larger the binary is, the slower and more memory-intensive its decompilation may be). Many cases are already reported in our original bug tracker, which we want to migrate into GitHub.

We plan to provide more information concerning these performance issues. Please, bear with us. There are a lot of things that still need to be done (fix of build issues on different systems, migration of our old bug tracker, wiki pages, documentation, tests, continuous integration, dependency management, roadmap, etc.). We try to do our best.

Until then, I would just like to point out that in most cases, these issues are NOT caused by a single, particular bug somewhere it the decompiler. Rather, improvements in multiple parts of the decompiler will be needed.

**Comment by Convery** [comment]
After 20 hours it crashed after having gone through 110GB of RAM: https://i.imgur.com/KgN8AR0.png

**Comment by Convery** [comment]
Yep, removing those from the `decompile.sh` as `--backend-no-opts` did nothing makes it finish in 90min and 10GB RAM.

![](https://i.imgur.com/VQWxqHu.png)

**Comment by mhsjlw** [comment]
It breaks at 2mb? I've got a binary that's 45 megabytes... what are my options? I've tried using `--backend-no-opts` and it'll just get stuck at `Running phase: Input binary to LLVM IR decoding ( 1.48s )`

**Comment by s3rvac** [comment]
@mhsjlw:

> It breaks at 2Mb?

It depends on the input binary. Many times, even larger binaries can be decompiled without a problem.

> I've got a binary that's 45 megabytes... what are my options?

Generally, when a decompilation takes too much time or consumes too much memory, you can perform a _selective decompilation_, which will decompile only a part of the input file. For example, by running
```
decompile.sh FILE --select-functions func1,func2 --select-decode-only
```
the decompiler will decode and decompile only functions named `func1` and `func2`. However, note that this works only when the input executable has symbol names in it. If it does not, you will have to specify address ranges to be decompiled via e.g.
```
decompile.sh FILE --select-ranges 0x404000-0x405000 --select-decode-only
```
This will decompile everything between addresses `0x404000` and `0x405000`.

> I've tried using --backend-no-opts and it'll just get stuck at Running phase: Input binary to LLVM IR decoding ( 1.48s )

This parameter only applies to the back-end phase (`llvmir2hll`). The phase you have mention is from the front-end part (`bin2llvmir`) + the decoding phase is mandatory as without decoding, the decompilation cannot go further.

**Comment by mhsjlw** [comment]
Ok, I'll give the selective decompilation a try, thanks.

**Comment by kiritowch** [comment]
I think you can try some selective decompilation...I’m not sure...

**Comment by Redict** [comment]
Same problem with another DLL. On Linux eats 16 GB RAM

**Comment by MerovingianByte** [comment]
Maybe the code could be refactored to use binary trees or some other memory efficient data structure.

**Comment by waldi** [comment]
Same for me 
![image](https://user-images.githubusercontent.com/11230/34979563-f48beece-faa1-11e7-9d25-3ce3a078a302.png)
<img width="776" alt="screen shot 2018-01-16 at 09 28 25" src="https://user-images.githubusercontent.com/11230/34979570-fa66580c-faa1-11e7-9844-78a432339631.png">
My Executable file is 16.4 MB big.

**Comment by tpruzina** [comment]
This is not a bug, it's a feature. I suspect that memory usage could be improved on case per case basis after some improvements to the logic are added, but shouting that some _random_ 10MB binary needs 10GB of RAM to decompile doesn't actually help anybody.

I suspect that if you open a separate issue write nice bug report (provide binary you have problem with along with decent debug/tracing/mem-usage info) or better yet suggest a patch, now that is something devs could work with.

I'm not sure where you people come from, but this is (hopefully) not how developers file bug reports. Perhaps this should be closed as INVALID?

**Comment by PeterMatula** [comment]
Few big files that we could test, and eventually handle.
[big-samples-1.zip](https://github.com/avast-tl/retdec/files/1673231/big-samples-1.zip)
[big-samples-2.zip](https://github.com/avast-tl/retdec/files/1673233/big-samples-2.zip)

**Comment by MerovingianByte** [comment]
@PeterMatula , you mean before this issue was opened or after?

**Comment by PeterMatula** [comment]
@MerovingianByte These are files from our internal issue-tracking system that I'm currently migrating here to GitHub. So they are from before this issue. The cause is mostly the same. We should eventually handle all the files from this issue -- both those I just added, and those from other users.

**Comment by MerovingianByte** [comment]
@PeterMatula I see. How much memory do those big files need?

**Comment by PeterMatula** [comment]
@MerovingianByte No idea :D At the time of reporting the original issue, enough to take down the decompilation process -- even whole system if per process limits are not set. Thanks to the changes in decompiler, some of them might be better now, but all of them should be checked anyway.

**Comment by MerovingianByte** [comment]
really looking forward to the fixing of all memory issues. I can't congratulate retdec enough. This project is just too good. The decompilation is more refined than hex-ray's. There's finally a viable, good, free alternative. Of course, there were other free alternatives before like snowman, but they were terrible. Is there an easy way to donate? If I was rich I'd be pouring money to this project. Like, shut up and take my money lol. But I think I can at least pay you a coffee.

**Comment by eMVe** [comment]
Same situation here.
Decompiling 9MB exe ate all of my 64GB RAM....
Firstly DeadLocalAssignOptimizer said: warning: out of memory: trying to recover
then SimpleCopyPropagationOptimizer crashed the windows...
I don't have swap file as my configuration is 60GB SSD and 64GB of RAM.

**Comment by Phrohdoh** [comment]
While this is indeed unfortunate, and I'm sure the retdec team is looking into it, users must deal with what we have available to use.

While analyzing the binary you feed it retdec creates a JSON document that describes the assembly (and the functions which is what we are mostly after). We can get function offsets and names from this.

To that end I have written a python 3 script that makes this somewhat easier (albeit still a bit tedious, I'm open to improvement suggestions).

```python
#!/usr/bin/python3

import os
import sys
import json

try:
    path_to_decompiler_sh = sys.argv[1]
    path_to_file_to_decompile = sys.argv[2]
    path_to_retdec_json = sys.argv[3]
except IndexError:
    print("Usage: path/to/decompiler.sh path/to/file.exe path/to/retdec.json")
    sys.exit(1)

with open(path_to_retdec_json, 'r') as f:
    obj = json.load(f)
    for fn in obj['functions']:
        n = fn['name']
        if not n.startswith("function_"):
            continue

        s = hex(fn['startAddr'])
        e = hex(fn['endAddr'])
        cmd = "{} {} --select-ranges {}-{} --select-decode-only -o {}.fn" \
            .format(path_to_decompiler_sh,
                    path_to_file_to_decompile,
                    s,
                    e,
                    n)

        print(cmd)
```

Running the script:

```sh
$ ./the-script.py ~/bin/decompiler.sh path/to/my.exe /path/to/retdec_output.json
```

Will produce something similar to the following:

```
/home/thill/bin/decompiler.sh path/to/my.exe --select-ranges 0x5350e9-0x5350e9 --select-decode-only -o function_5350e9.fn
/home/thill/bin/decompiler.sh path/to/my.exe --select-ranges 0x53fbd8-0x53fbd8 --select-decode-only -o function_0x53fbd8.fn
```

Which I can then run directly.

You can of course modify the script to use `subprocess`/`popen`.

Hopefully this helps someone.

**Comment by 0xBEEEF** [comment]
I would also like to add a little more. When optimizing in the backend, I mainly encounter problems with the SimpleCopyPropagationOptimizer. If I exclude it, I have no problems with most examples and tests. So there seems to be some mistake here.

**Comment by rkalz** [comment]
@0xBEEEF What flag did you use/remove to disable that step?

**Comment by PeterMatula** [comment]
Sample from #97 is also taking to much time and memory in `llvmir2hll`.

**Comment by PeterMatula** [comment]
No, just zip it and upload it here if you want to share it for testing.

**Comment by MPeti1** [comment]
Ok, thanks, [here](https://1drv.ms/f/s!ArvLKxtjzZnrlqpiNLea74HG0TkrTw) is the link.
Please note that this file is somewhat like the op's file in terms of availability, and that it's an android app's library working as a [native Activity](https://developer.android.com/reference/android/app/NativeActivity)

**Comment by twifty** [comment]
I had a similar problem with a 1.2mb (not packed) file. DeadLocalAssignOptimizer issued "warning: out of memory: trying to recover" in a seemingly endless loop. Though only 17GB of a total 32GB RAM was actually used.

Adding the `--backend-no-opts` arg and the whole process completed in seconds.

**Comment by denji** [comment]
`retdec-decompiler.py --backend-disabled-opts DeadCode,DeadLocalAssign` LGTM

**Comment by denji** [comment]
`env LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2 retdec-decompiler.py --backend-disabled-opts=DeadCode,DeadLocalAssign --no-memory-limit ./app.bin` LGTM
* [`dead_code_optimizer`](https://github.com/avast/retdec/blob/v3.3/include/retdec/llvmir2hll/optimizer/optimizers/dead_code_optimizer.h#L113)
* [`dead_local_assign_optimizer`](https://github.com/avast/retdec/blob/v3.3/include/retdec/llvmir2hll/optimizer/optimizers/dead_local_assign_optimizer.h#L47)

**Comment by denji** [comment]
```sh
retdec-decompiler.py --backend-disabled-opts=DeadCode,DeadLocalAssign,CopyPropagation,SimpleCopyPropagation ...
```

**Comment by Mortrus** [comment]
I have same problem. I'm trying to decompile some.dll (11 Mb), and now I stuck for 20 hours at this step with enormous memory consumption:
![Безымянный](https://user-images.githubusercontent.com/42960438/63646038-f2f2d600-c72c-11e9-9fdf-1177609c2dc2.png)

Decompilation settings are all default. Just set no memory limit. -_-

**Comment by sooxt98** [comment]
![image](https://user-images.githubusercontent.com/13378059/83656578-4e25f300-a5f2-11ea-9819-1e1878e8a264.png)

same here, it just stop somewhere using half of my ram

**Comment by rmast** [comment]
I finally decided to rent an Azure-server to be able to allow as much optimization as possible using virtual memory.
I selected the default batch-runner Windows 2016 server image on a machine with 2 cores, 64GB RAM and 128GB SSD + 128GB TEMP disk. I added another 400GB SSD, partitioned, formatted and assigned it a drive letter, but that appeared not to be necessary as the decompilation until now hasn't taken more than about 100GB committed memory in the task overview.

I upped the swap-file to those disks outside C: to max-size, however enabling system managed swap on enough disks would already allow swap 3 times RAM = 192 GB,

I allowed IE usage, installed python 3.7.3 on the server, ran retdec-decompiler.py with the --no-memory-limit option, turned off the automatic turn off of the server at 7PM UTC, turned off Windows update that rebooted my system around 6 AM, and even installed Application Insights to follow the committed memory usage from outside the server without having to RDP to it.

At the moment the server decompile-process is finally surviving the night.

**Comment by SamDecrock** [comment]
As a reference, I once tried to decompile a 30MB binary. It ended up using 800GB of RAM 😮

**Comment by kiritowch** [comment]
这是一封自动回复邮件。已经收到您的来信，我会尽快回复。
